diff --git a/archival/libarchive/open_transformer.c b/archival/libarchive/open_transformer.c
index 44715ef..0dcf231 100644
--- a/archival/libarchive/open_transformer.c
+++ b/archival/libarchive/open_transformer.c
@@ -189,6 +189,13 @@ static transformer_state_t *setup_transformer_on_fd(int fd, int fail_if_not_comp
 		USE_FOR_NOMMU(xstate->xformer_prog = "bunzip2";)
 		goto found_magic;
 	}
+	if (ENABLE_FEATURE_SEAMLESS_LZMA
+	 && xstate->magic.b16[0] == LZMA_MAGIC
+	) {
+		xstate->xformer = unpack_lzma_stream;
+		USE_FOR_NOMMU(xstate->xformer_prog = "unlzma";)
+		goto found_magic;
+	}
 	if (ENABLE_FEATURE_SEAMLESS_XZ
 	 && xstate->magic.b16[0] == XZ_MAGIC1
 	) {
@@ -207,6 +214,7 @@ static transformer_state_t *setup_transformer_on_fd(int fd, int fail_if_not_comp
 	if (fail_if_not_compressed)
 		bb_simple_error_msg_and_die("no gzip"
 			IF_FEATURE_SEAMLESS_BZ2("/bzip2")
+			IF_FEATURE_SEAMLESS_LZMA("/lzma")
 			IF_FEATURE_SEAMLESS_XZ("/xz")
 			" magic");
 
diff --git a/include/bb_archive.h b/include/bb_archive.h
index e0ef8fc..3e54655 100644
--- a/include/bb_archive.h
+++ b/include/bb_archive.h
@@ -9,6 +9,7 @@ enum {
 	COMPRESS_MAGIC = 0x1f9d,
 	GZIP_MAGIC  = 0x1f8b,
 	BZIP2_MAGIC = 256 * 'B' + 'Z',
+	LZMA_MAGIC  = 0x5d00,
 	/* .xz signature: 0xfd, '7', 'z', 'X', 'Z', 0x00 */
 	/* More info at: http://tukaani.org/xz/xz-file-format.txt */
 	XZ_MAGIC1   = 256 * 0xfd + '7',
@@ -21,6 +22,7 @@ enum {
 	COMPRESS_MAGIC = 0x9d1f,
 	GZIP_MAGIC  = 0x8b1f,
 	BZIP2_MAGIC = 'B' + 'Z' * 256,
+	LZMA_MAGIC  = 0x5d,
 	XZ_MAGIC1   = 0xfd + '7' * 256,
 	XZ_MAGIC2   = 'z' + ('X' + ('Z' + 0 * 256) * 256) * 256,
 	XZ_MAGIC1a  = 0xfd + ('7' + ('z' + 'X' * 256) * 256) * 256,
